src/EditorFeatures/Core/Implementation/Classification/SyntacticClassificationTaggerProvider.TagComputer.cs

```diff
@@ -33,342 +33,409 @@ internal partial class SyntacticClassificationTaggerProvider
-        internal partial class TagComputer
+        internal partial class TagComputer : ForegroundThreadAffinitizedObject
         {
+            private readonly SyntacticClassificationTaggerProvider _taggerProvider;
             private readonly ITextBuffer _subjectBuffer;
-            private readonly WorkspaceRegistration _workspaceRegistration;
-            private readonly AsynchronousSerialWorkQueue _workQueue;
+            private readonly IAsynchronousOperationListener _listener;
+            private readonly ClassificationTypeMap _typeMap;
 
-            // this will cache previous classification information for a span, so that we can avoid
-            // digging into same tree again and again to find exactly same answer
-            private readonly LastLineCache _lastLineCache;
+            private readonly WorkspaceRegistration _workspaceRegistration;
 
-            // The latest data about the document being classified that we've cached.  objects can 
-            // be accessed from both threads, and must be obtained when this lock is held.
-            private readonly object _gate = new();
-            private ITextSnapshot _lastProcessedSnapshot;
-            private Document _lastProcessedDocument;
+            private readonly CancellationTokenSource _disposalCancellationSource = new();
 
             /// <summary>
-            /// Data stored by the <see cref="IClassificationService"/> that can be computed in the bg which it will
-            /// benefit from when it is called back for classifications later.
+            /// Work queue we use to batch up notifications about changes that will cause
+            /// us to classify.  This ensures that if we hear a flurry of changes, we don't
+            /// kick off an excessive amount of background work to process them.
             /// </summary>
-            private object _lastProcessedCachedData;
-
-            private Workspace _workspace;
-            private readonly CancellationTokenSource _reportChangeCancellationSource = new();
-
-            private readonly IAsynchronousOperationListener _listener;
-            private readonly IForegroundNotificationService _notificationService;
-            private readonly ClassificationTypeMap _typeMap;
-            private readonly SyntacticClassificationTaggerProvider _taggerProvider;
+            private readonly AsyncBatchingWorkQueue<ITextSnapshot> _workQueue;
 
             /// <summary>
             /// Timeout before we cancel the work to diff and return whatever we have.
             /// </summary>
             private readonly TimeSpan _diffTimeout;
 
+            private Workspace? _workspace;
+
+            // The latest data about the document being classified that we've cached.  objects can 
+            // be accessed from both threads, and must be obtained when this lock is held. 
+            //
+            // _lastProcessedRoot is optional and is held onto for languages that support syntax.
```

Comment:
?
```suggestion
            // _lastProcessedData is optional and is held onto for languages that support syntax.
```

Reply:
ah, slightly stale comment.  it should be _lastProcessedData.lastRoot